### PR TITLE
model: dynamic output class

### PR DESCRIPTION
**Pull request by rebel-jongho**
# Pull Request Description

## Type of Change

- [ ] New Model Support
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring
- [ ] Other (please describe):

## Motivation and Context

In the old code, the forward function of a model class with multiple outputs would assign multiple outputs to a single field in BaseModelOutput.

This PR fixes that.

However, there is a case where returning a tuple is still behaving differently than huggingface (flattening nested tensors), which will need to be fixed in the future.

## Checklist

- [x] I have performed a self-review of my own code
- [ ] I have added tests that prove my fix is effective or that my feature works (If needed)

## Related Issues

----
# Conventional commit
```
type(optional scope): description
```
----
# Type candidate
  - Model Updates
    - `model`: Adding New models or Bugfix for existing models
      - ex) Add LlavaNext 
      - ex) Bugfix Whisper
  - Enhancements
    - `performance`: Optimizing some models or this library itself
      - ex) Loading RBLNModel faster
      - ex) Optimizing Memory Usage of DecoderOnlyModel
  - Code Refactor
    - `refactor`: Re-arrange class architecture, or more.
      - ex) Refactor Seq2Seq
  - Documentation
    - `doc`: Update docstring only     
  - Library Dependencies
    - `dependency`: Update requirements, something like that.
  - Other
    - `other`: None of above.
      - ex) ci update
      - ex) pdm update

### CODE REVIEW

File: src/optimum/rbln/modeling.py [hunk 1]
```diff
@@ -245,16 +244,66 @@ def forward(self, *args, return_dict: Optional[bool] = None, **kwargs):
...
+                                    return cls._output_class
+                    else:
+                        if hasattr(return_type, "__module__") and (
+                            "transformers" in return_type.__module__ or "diffusers" in return_type.__module__
+                        ):
+                            cls._output_class = return_type
+                            return cls._output_class
+            else:
+                # Fallback to BaseModelOutput
+                cls._output_class = BaseModelOutput
```
Comment (rebel-kblee) [inline src/optimum/rbln/modeling.py:280]: If there is no forward method, it always maps to BaseModelOutput, but I'm guessing there must be an exception case?

File: src/optimum/rbln/modeling.py [hunk 2]
```diff
@@ -245,16 +244,66 @@ def forward(self, *args, return_dict: Optional[bool] = None, **kwargs):
...
-            return (output,) if not isinstance(output, (tuple, list)) else output
+            return tuple_output
         else:
-            if self.output_class is None:
-                return BaseModelOutput(last_hidden_state=output)
-
-            # Create output with the appropriate class and key
-            return self.output_class(**{self.output_key: output})
+            output_class = self.get_hf_output_class()
+            if hasattr(output_class, "loss"):
```
Comment (rebel-kblee) [inline src/optimum/rbln/modeling.py:295]: only for causalLM? What about other cases that might not be related to loss?

File: src/optimum/rbln/modeling.py (hunk 1 above)
Comment (rebel-jongho) [reply src/optimum/rbln/modeling.py:280]: The `forward` method always exists Since transformer's PretrainedModel / diffusers model classes inherits from torch.nn.Module.

Comment (rebel-thkim) [reply src/optimum/rbln/modeling.py:280]: In the case of inference using the `__call__ ` method, is it possible that (diffuser) will never encounter this logic?

**Review by rebel-thkim — APPROVED** [review]

File: src/optimum/rbln/modeling.py (hunk 2 above)
Comment (rebel-jongho) [reply src/optimum/rbln/modeling.py:295]: this is not only for causal lm. why did you think that?

File: src/optimum/rbln/modeling.py (hunk 1 above)
Comment (rebel-jongho) [reply src/optimum/rbln/modeling.py:280]: In case of diffusers, they use `RBLNDiffuserMixin` (which uses `__call__`) for pipelines and RBLNModel(which uses `forward`) for its submodules. 

So, you are right, diffusers classes will never encounter this logic.

Comment (rebel-junamsong) [reply src/optimum/rbln/modeling.py:280]: @rebel-jongho How about writing it like this?

```
    @classmethod
    def get_hf_output_class(cls):
        if cls._output_class:
            return cls._output_class

        hf_class = cls.get_hf_class()
        if hf_class is None:
            raise ValueError(f"No HuggingFace model class found for {cls.__name__}")

        hints = get_type_hints(hf_class.forward) if hasattr(hf_class, "forward") else {}
        ret = hints.get("return")
        candidates = get_args(ret) if get_origin(ret) is Union else (ret,)

        for t in candidates:
            mod = getattr(t, "__module__", "")
            if "transformers" in mod or "diffusers" in mod:
                cls._output_class = t
                return t

        cls._output_class = BaseModelOutput
        return BaseModelOutput
```

File: src/optimum/rbln/modeling.py (hunk 2 above)
Comment (rebel-kblee) [reply src/optimum/rbln/modeling.py:295]: if output_class == BaseModelOutput, BaseModelOutput has no attribute "loss". 
With this case, It does not add for tuple_output

File: src/optimum/rbln/modeling.py (hunk 1 above)
Comment (rebel-jongho) [reply src/optimum/rbln/modeling.py]: @rebel-junamsong I apllied it, with additional changes. Thanks
